### PR TITLE
Add SIMD accelerated map_boxes

### PR DESCRIPTION
**Pull request by April-Hannah-Lena**
This PR add a new elegant SIMD-accelerated version of `map_boxes` which runs between 3x and 5x faster than the standard version. It works completely automatically, you can use your regular function and simply pass `:cpu` arg to `BoxMap` (More details in `fast_maps_using_SIMD.jl`).

```julia
function map_boxes(g::SampledBoxMap{N,T,Val{:cpu}}, source::BoxSet) where {N,T}
    P, keys = source.partition, collect(source.set)
    image = fill( Set{eltype(keys)}(), nthreads() )
    test_points = g.domain_points(P.domain.center, P.domain.radius)
    n, simd = length(test_points), pick_vector_width(T)
    @threads for key in keys
        box = key_to_box(P, key)
        c, r = box.center, box.radius
        for i in 0:simd:n-simd
            points = tuple_vgather(view(test_points, i+1:i+simd), simd)
            points = @muladd points .* r .+ c
            points = g.map(points)
            keys = point_to_key(P, points)
            push!(image[threadid()], keys...)
        end
    end
    return BoxSet(P, union(image...))
end
```

I realized that I was looking to speed up the wrong part of `map_boxes` in my previous attempts. I've always been trying to parallelize the function itself, inevitably needing more allocations. However, what I really needed to do was parallelize the test points and `point_to_key` functions. 

The work was worth it, since I was able to reuse much of my old code, just in new places. To run a benchmark, clone April-Hannah-Lena/GAIO.jl@55423eaf84b5646d8827ab26092524c9d771a068 and run

```julia
julia> cd("./benchmark")

julia> include("map_boxes_bench.jl")

Standard Version
G = BoxMap(F, P) = BoxMap with 48 sample points
W = unstable_set!(G, P[x]) = 18305-element BoxSet in 3 dimensions
  0.354856 seconds (9.71 M allocations: 474.702 MiB, 30.61% gc time)
  0.298456 seconds (9.70 M allocations: 474.534 MiB, 17.57% gc time)
  0.299194 seconds (9.70 M allocations: 474.802 MiB, 16.45% gc time)
  0.301404 seconds (9.70 M allocations: 474.284 MiB, 16.97% gc time)
  0.299297 seconds (9.71 M allocations: 474.727 MiB, 16.55% gc time)
  0.297946 seconds (9.70 M allocations: 474.649 MiB, 16.37% gc time)
  0.295239 seconds (9.70 M allocations: 474.701 MiB, 16.55% gc time)
  0.297496 seconds (9.69 M allocations: 474.218 MiB, 16.53% gc time)
  0.301596 seconds (9.70 M allocations: 474.644 MiB, 16.29% gc time)
  0.341864 seconds (9.70 M allocations: 474.764 MiB, 29.66% gc time)

cpu version
G = BoxMap(F, P, :cpu) = BoxMap with 48 sample points
W = unstable_set!(G, P[x]) = 19193-element BoxSet in 3 dimensions
  0.091189 seconds (4.38 M allocations: 212.037 MiB)
  0.145110 seconds (4.38 M allocations: 211.709 MiB, 36.48% gc time)
  0.090975 seconds (4.38 M allocations: 212.038 MiB)
  0.141194 seconds (4.38 M allocations: 211.950 MiB, 36.48% gc time)
  0.093089 seconds (4.38 M allocations: 211.976 MiB)
  0.137966 seconds (4.38 M allocations: 211.783 MiB, 33.45% gc time)
  0.096850 seconds (4.38 M allocations: 212.095 MiB)
  0.138624 seconds (4.38 M allocations: 211.709 MiB, 33.46% gc time)
  0.092992 seconds (4.38 M allocations: 211.900 MiB)
  0.137625 seconds (4.38 M allocations: 211.901 MiB, 33.99% gc time)

```

This PR also moves away from `StaticArays` in favor of tuples. I'm not fully sure if this is worth the speed benefit since we lose many of the automatic `AbstractArray` methods. All the examples and tests work (some examples need to manually import `StaticArrays`). I think this is something we could discuss later if you want.

### CODE REVIEW

**Comment by gaioguy** [comment]
Ok, very nice. I get
```
julia> G = BoxMap(F, P, no_of_points=400, :cpu)
julia> @benchmark G($B)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):   83.541 μs …  11.899 ms  ┊ GC (min … max): 0.00% … 98.71%
 Time  (median):     102.094 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   113.379 μs ± 288.272 μs  ┊ GC (mean ± σ):  6.13% ±  2.42%

          █                                                      
  ▁▁▁▁▁▂▃▆█▇▇▄▂▂▂▂▂▂▂▂▂▃▄▄▅▆▅▄▃▃▃▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  83.5 μs          Histogram: frequency by time          150 μs <

 Memory estimate: 90.78 KiB, allocs estimate: 2427.

julia> G = BoxMap(F, P, no_of_points=400)
BoxMap with 400 sample points

julia> @benchmark G($B)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  257.975 μs …  13.342 ms  ┊ GC (min … max): 0.00% … 97.51%
 Time  (median):     291.707 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   314.617 μs ± 524.931 μs  ┊ GC (mean ± σ):  6.82% ±  4.01%

   ▃   █                                                         
  ███▅▇█▇▆▄▃▃▃▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▂▁▂▂▁▁▂ ▃
  258 μs           Histogram: frequency by time          673 μs <

 Memory estimate: 220.80 KiB, allocs estimate: 4422.
```
while with the current master branch, I get
```
julia> F = BoxMap(f, P, no_of_points=400)
BoxMap with 400 sample points

julia> @benchmark F($B)
BenchmarkTools.Trial: 
  memory estimate:  1.97 KiB
  allocs estimate:  20
  --------------
  minimum time:     194.639 μs (0.00% GC)
  median time:      207.646 μs (0.00% GC)
  mean time:        214.279 μs (0.00% GC)
  maximum time:     953.428 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```
So it actually is a nice improvement! We might still want to think about the larger number of allocations at some point.

**Comment by April-Hannah-Lena** [comment]
Ok I rebased with the current master branch, so the only changes are to the :cpu BoxMap. The normal one is now untouched. 

I also was able to squeeze out another 8% speed by moving the SIMD gather operation outside of the loop.

```julia
function map_boxes(g::SampledBoxMap{N,T,Val{:cpu}}, source::BoxSet) where {N,T}
    P, keys = source.partition, collect(source.set)
    image = [ Set{eltype(keys)}() for _ in  1:nthreads() ]
    points = tuple_vgather(
        g.domain_points(P.domain.center, P.domain.radius),
        pick_vector_width(T)
    )
    @threads for key in keys
        box = key_to_box(P, key)
        c, r = box.center, box.radius
        for p in points
            fp = g.map(@muladd p .* r .+ c)
            hits = point_to_key(P, fp)
            push!(image[threadid()], hits...)
        end
    end
    return BoxSet(P, union(image...))
end
```

**Comment by gaioguy** [comment]
I am getting quite mixed results, depending on the number of sample points:
the current master
```julia
julia> G = BoxMap(F, P, no_of_points=40)
julia> @time W = unstable_set!(G, P[x])
  0.066494 seconds (5.65 k allocations: 4.212 MiB)
```
vs. with the PR
```julia
julia> G = BoxMap(F, P, :cpu, no_of_points=40)
julia> @time W = unstable_set!(G, P[x])
  0.252924 seconds (3.76 M allocations: 147.461 MiB)
``` 
and
```julia
julia> G = BoxMap(F, P, no_of_points=4000)
julia> @time W = unstable_set!(G, P[x])
 23.375448 seconds (8.63 k allocations: 13.036 MiB)
```
vs. the PR
```julia
julia> G = BoxMap(F, P, :cpu, no_of_points=4000)
julia> @time W = unstable_set!(G, P[x])
 14.729254 seconds (617.75 M allocations: 22.918 GiB, 30.34% gc time)
```
and
```julia
julia> G = BoxMap(F, P, no_of_points=40000)
julia> @time W = unstable_set!(G, P[x])
248.886638 seconds (8.39 k allocations: 13.699 MiB)
```
vs. the PR
```julia
julia> G = BoxMap(F, P, :cpu, no_of_points=40000)
julia> @time W = unstable_set!(G, P[x])
284.639040 seconds (6.39 G allocations: 236.881 GiB, 39.71% gc time)
``` 
In fact, with 40000 points, the accelerated code is doing a huge number of allocations, while the number of allocations for the current code in the master is actually independent of the number of sample points. I guess we still need to work on the allocations issue ...

File: src/algorithms.jl [hunk 1]
```diff
@@ -95,27 +95,32 @@ function finite_time_lyapunov_exponents(boxset::BoxSet, g; T, num_points=20, ϵ=
...
 
-    k = f(@. x+τ*k)
-    dx = @. dx + k/6
+    k  = f(k)
+    dx = @. dx + k * sixth
+    k  = @. x + τ * dx
 
-    return @. x+τ*dx
+    return k
 end
```
Comment (gaioguy) [inline src/algorithms.jl:119]: How much do we gain be re-writing it this way?

File: src/boxmap.jl [hunk 2]
```diff
@@ -1,42 +1,65 @@
...
     domain::Box{N,T}
-    domain_points::P
-    image_points::I    
+    domain_points
+    image_points
+    acceleration::B
 end
 
-function Base.show(io::IO, g::SampledBoxMap)
+function Base.show(io::IO, g::BoxMap)
```
Comment (gaioguy) [inline src/boxmap.jl:28]: I guess we really want `SampledBoxMap` here, since otherwise the number of sample points might not be defined.

File: src/boxmap.jl [hunk 3]
```diff
@@ -55,37 +78,49 @@ end
 
 function AdaptiveBoxMap(f, domain::Box{N,T}) where {N,T}
     Df = x -> ForwardDiff.jacobian(f, x)
-    domain_points = (center, radius) -> sample_adaptive(Df, center)
+    domain_points(center, radius) = sample_adaptive(Df, center)
```
Comment (gaioguy) [inline src/boxmap.jl:81]: Is that speeding up things or rather a cosmetic change?

File: src/boxmap.jl [hunk 4]
```diff
@@ -55,37 +78,49 @@ end
...
+    P, keys = source.partition, collect(source.set)
+    image = [ Set{eltype(keys)}() for _ in  1:nthreads() ]
+    points = g.domain_points(P.domain.center, P.domain.radius)
+    @threads for key in keys
+        box = key_to_box(P, key)
+        c, r = box.center, box.radius
+        for p in points
+            fp = g.map(@muladd p .* r .+ c)
+            hits = point_to_key(P, fp)
+            push!(image[threadid()], hits...)
```
Comment (gaioguy) [inline src/boxmap.jl:120]: Don't we need to check whether `hits` is `nothing`?

File: src/partition_regular.jl [hunk 5]
```diff
@@ -29,7 +29,7 @@ end
 function BoxPartition(domain::Box{N,T}, dims::NTuple{N,Int}) where {N,T}
     dims = SVector{N,Int}(dims)
     left = domain.center .- domain.radius
-    scale = dims ./ (2*domain.radius)
+    scale = dims ./ (2 .* domain.radius)
```
Comment (gaioguy) [inline src/partition_regular.jl:32]: Can we actually move these rather cosmetic changes into a separate PR? The smaller the PR, the better ...

**Comment by April-Hannah-Lena** [comment]
So I took out all changes inclusing cosmetic changes in order to track down these allocations, but for some reason I smiply can't recreate those results. In April-Hannah-Lena/GAIO.jl@9f27aa7f7d6c882674276c388c00b2df61c97804 I only added code for the :cpu version, did not change any existing code, and did a benchmark:

```julia
using GAIO

# the Lorenz system
const σ, ρ, β = 10.0, 28.0, 0.4
v((x,y,z)) = (σ*(y-x), ρ*x-y-x*z, x*y-β*z)
f(x) = rk4_flow_map(v, x)

center, radius = (0,0,25), (30,30,30)
x = (sqrt(β*(ρ-1)), sqrt(β*(ρ-1)), ρ-1)         # equilibrium
P = BoxPartition(Box(center, radius), (128,128,128))

for accel in (nothing, :cpu), p in (40, 400, 4000)
    println("accel: $accel  p: $p")
    F = BoxMap(f, P, accel; no_of_points=p)
    W = unstable_set!(F, P[x])
    @time W = unstable_set!(F, P[x])
end
```

I get the follownig results:

```
accel: nothing  p: 40
  0.196086 seconds (8.56 M allocations: 436.467 MiB)
accel: nothing  p: 400
  4.601838 seconds (135.84 M allocations: 6.703 GiB, 41.57% gc time)
accel: nothing  p: 4000
 52.546914 seconds (1.59 G allocations: 75.005 GiB, 44.91% gc time)
accel: cpu  p: 40
  0.051536 seconds (2.96 M allocations: 122.944 MiB)
accel: cpu  p: 400
  1.201683 seconds (53.17 M allocations: 2.083 GiB, 39.96% gc time)
accel: cpu  p: 4000
 15.361004 seconds (598.18 M allocations: 23.355 GiB, 48.56% gc time)
```

The allocations also only seem to come from mapping the boxes and point_to_key, so it's the operations that are repeated in linear proportion to the number of test points / number of keys. The number of allocations is always cut by a factor of 2-4, but are never constant. 

```
 ────────────────────────────────────────────────────────────────────────────────────────────
                                                    Time                    Allocations      
                                           ───────────────────────   ────────────────────────
             Tot / % measured:                  6.06s /  22.0%            803MiB /   4.0%    

 Section                           ncalls     time    %tot     avg     alloc    %tot      avg
 ────────────────────────────────────────────────────────────────────────────────────────────
 BoxMap with 40000 sample points        1    1.06s   79.4%   1.06s   22.0MiB   68.9%  22.0MiB
   map                              40.0k    1.01s   75.7%  25.2μs   11.0MiB   34.4%     288B
     rk4_flow_map                   40.0k    979ms   73.3%  24.5μs      752B    0.0%    0.02B
       lorenz                       3.20M    109ms    8.2%  34.2ns     0.00B    0.0%    0.00B
   point_to_key                     40.0k   6.80ms    0.5%   170ns   6.71MiB   21.0%     176B
   push!                            40.0k   3.40ms    0.3%  85.0ns    625KiB    1.9%    16.0B
   key_to_box                           1    554ns    0.0%   554ns     0.00B    0.0%    0.00B
 BoxMap with 10000 sample points        1    275ms   20.6%   275ms   9.94MiB   31.1%  9.94MiB
   map                              10.0k    239ms   17.9%  23.9μs   2.14MiB    6.7%     224B
     rk4_flow_map                   10.0k    236ms   17.7%  23.6μs      752B    0.0%    0.08B
       lorenz                        800k   26.4ms    2.0%  33.0ns     0.00B    0.0%    0.00B
   point_to_key                     10.0k   22.6ms    1.7%  2.26μs   5.95MiB   18.6%     624B
   push!                            10.0k   3.66ms    0.3%   366ns   1.83MiB    5.7%     192B
   key_to_box                           1    789ns    0.0%   789ns     0.00B    0.0%    0.00B
 ────────────────────────────────────────────────────────────────────────────────────────────
```

**Comment by April-Hannah-Lena** [comment]
Nevermind, after tearing my hair out I found the issue. Results now look like this:

```
accel: nothing  p: 40
  0.023969 seconds (8.90 k allocations: 4.056 MiB)
accel: nothing  p: 400
  0.567254 seconds (14.04 k allocations: 9.345 MiB)
accel: nothing  p: 4000
  5.770438 seconds (14.26 k allocations: 11.184 MiB)
accel: cpu  p: 40
  0.018903 seconds (12.97 k allocations: 5.337 MiB)
accel: cpu  p: 400
  0.229319 seconds (14.04 k allocations: 9.357 MiB)
accel: cpu  p: 4000
  2.463823 seconds (14.29 k allocations: 11.113 MiB)
```

**Comment by gaioguy** [comment]
Indeed, in various tests, even without acceleration I already get a 2x speedup over the current master and with `:cpu` acceleration even a 4x speedup. Awesome ...
